### ISSUE TITLE
Composer uses excessive relative path notation "../" in conjunction with $baseDir in autoload_psr4.php (only in certain environments)

### ISSUE THREAD

**Issue by mmic-bjohnson** [composer/composer#4387]
This issue seems similar to https://github.com/composer/composer/issues/3024 . Hopefully, there is a rational explanation for it and a simple configuration change will suffice.

I have several similarly-configured systems in that they all use the same OS, web-stack components, and software versions. The only difference between these systems (at least that I'm able to discern) is that on the system that does *not* exhibit this problem, the `vendor` directory exists directly inside the web-server's document root, whereas on the problematic systems, the `vendor` directory is up one level from the web-server's document root.

The basic nature of the problem is that the paths that are used in conjunction with `$baseDir` in Composer's `autoload_psr4.php` contain seemingly excessive relative path notation. For whatever reason, the problem only occurs with `$baseDir` and *not* `$vendorDir`. To be clear, classes are still auto-loaded correctly; I'm just wondering why all the extra `../`. And more importantly, I'd like to eliminate these extra `../`'s because they are complicating my build process (for reasons similar to those explained in the Issue cited above).

The `composer.json` is identical on all three systems. I'm happy to post the full contents if it is relevant:

```json
{
	"autoload": {
		"classmap": [
			"database"
		],
		"files": [
			"app/functions.php"
		],
		"psr-4": {
			"App\\": "app/",
			"Cartalyst\\Platform\\": "app/overrides/src"
		}
	}
}
```

On the unaffected system, the entries look to be correct:

```php
return array(
//...
    'Platform\\Users\\' => array($baseDir . '/extensions/platform/users/src'),
    'Platform\\Themes\\' => array($baseDir . '/extensions/platform/themes/src'),
    'Platform\\Tags\\' => array($baseDir . '/extensions/platform/tags/src'),
//...
);
```

If I stick `var_dump(realpath($vendorDir), realpath($baseDir));exit;` on line 7 of `autoload_psr4.php`, the values are:

`string(40) "/var/www/clients/client1/web2/web/vendor" string(33) "/var/www/clients/client1/web2/web" `

By contrast, on one of the affected systems, the entries have redundant `../`'s:

```php
return array(
//...
    'Platform\\Users\\' => array($baseDir . '/../../../../../sub.example.com/private/laravel/extensions/platform/users/src'),
    'Platform\\Themes\\' => array($baseDir . '/../../../../../sub.example.com/private/laravel/extensions/platform/themes/src'),
    'Platform\\Tags\\' => array($baseDir . '/../../../../../sub.example.com/private/laravel/extensions/platform/tags/src'),
//...
);
```

Again, if I stick `var_dump(realpath($vendorDir), realpath($baseDir));exit;` on line 7 of `autoload_psr4.php`, the values are:

`string(52) "/var/www/clients/client1/web5/private/laravel/vendor" string(45) "/var/www/clients/client1/web5/private/laravel"`

My understanding (based on the Issue cited above) is that Composer should resolve these paths relative to the `composer.json` file. Given that in the second instance, `composer.json` resides in `/var/www/clients/client1/web5/private/laravel`, shouldn't these entries look just like the ones on the unaffected system?

In other words, couldn't the `/../../../../../sub.example.com/private/laravel` be removed from every entry without affecting the behavior? And in fact, if I replace `/../../../../../sub.example.com/private/laravel` by an empty string, throughout the file, everything still works.

Given that I am already using relative autoload paths in `composer.json`, what might explain this behavior?

Thank you for any assistance with this.

**Comment by stof** [comment]
Can you also show the definitions of the baseDir and vendorDir variables, which should be at the beginning of the file, before the array ?

**Comment by mmic-bjohnson** [comment]
Hello stof, and thank you for the incredibly prompt reply. I really appreciate your time.

The definitions are identical in all environments:

```php
$vendorDir = dirname(dirname(__FILE__));
$baseDir = dirname($vendorDir);
```

Please let me know if I can provide any additional information.

**Comment by stof** [comment]
@mmic-bjohnson Is ``sub.example.com`` as symlink for ``web5`` ? This might be the cause of the issue.
Can you try dumping the autoload while being in the real location of the project rather than being in a symlinked location ?

**Comment by mmic-bjohnson** [comment]
@stof Yes, you're exactly correct: `/var/www/sub.example.com/` is a symlink to `/var/www/clients/client1/web5`.

I tried `cd`ing into `/var/www/clients/client1/web5` before executing `composer dumpautoload`, but the results appear to be the same.

**Comment by mmic-bjohnson** [comment]
@stof I just remembered that I had to change the paths in Laravel's `bootstrap/paths.php` to get this working correctly in the first place. Surely, this is part of the problem.

I changed

```php
return [
	'base'   => '/../../../..',
	'public' => '/../../../../../../web',
];
```

to

```php
return [
	'base'   => '/var/www/sub.example.com/private/laravel',
	'public' => '/var/www/sub.example.com/web',
];
```

I hard-coded those paths only because, for whatever reason, I simply could not get it to work with any formulation of relative paths.

Any idea how I might modify the above paths to work correctly for my environment?

Thanks again.

**Comment by mmic-bjohnson** [comment]
After posting the above comment, I thought to myself, "Actually, that's silly. Why would Composer care about some Laravel-specific bootstrap directive? This value can't be a contributing factor."

But when I change the paths to to "relative to `composer.json`"

```php
return [
	'base'   => '',
	'public' => '/../../web',
];
```

the paths appear to be correct in `autoload_psr4.php`:

```php
return array(
    //...
    'Platform\\Users\\' => array($baseDir . '/extensions/platform/users/src'),
    'Platform\\Themes\\' => array($baseDir . '/extensions/platform/themes/src'),
    'Platform\\Tags\\' => array($baseDir . '/extensions/platform/tags/src'),
    //...
);
```

Unfortunately, this causes a failure with a third-party library:

```
 [RuntimeException]
  /../../web/themes/admin/default does not exist and could not be created.
```

Maybe the problem is actually with the third-party library -- and in particular how it resolves relative paths -- and not with Composer. That might explain why using the absolute path works in both places (despite all the extra `../` instances). Maybe this third-party library is assuming that this path is relative to a different directory than Composer is assuming.

This leaves just one question: why does a Laravel bootstrap file influence how Composer generates its PSR-4 autoload paths? Or am I misunderstanding something fundamental?

Thanks again for taking a look at this.

**Comment by mmic-bjohnson** [comment]
Upon digging a bit deeper, it looks like this is not a Composer issue.

I notice that only the autoload paths for this specific third-party library are a) subject to these extra `../` and b) influenced by this Laravel-specific bootstrap directive.

I don't understand all the details (and they're not important), but from what I'm able to determine, this third-party library hooks into Composer to perform some aspect of its installation and in the process poisons the `$vendorDir` and `$baseDir` values, thus leading to this issue.

I will take it up with the other library.

Thanks a lot for your time @stof and I apologize for wasting it!

**Comment by stof** [comment]
@mmic-bjohnson can you tell us which library it is ? This would help if someone else faces such issues in the future.

**Comment by stof** [comment]
btw, if the other library expects a relative path, you should remove the leading ``/`` from your path. Something starting with a ``/`` is *not* a relative path, but an absolute one

**Comment by mmic-bjohnson** [comment]
@stof Great spot! Removing the leading slash actually seems to have fixed the problem altogether!

The correct values seem to be:

```php
return [
	'base'   => './',
	'public' => '../../web',
];
```

I had assumed (quite erroneously) that this `../../web` string would be appended to some "base path" that does not end with a slash, e.g., `$path = '/var/www/sub.example.com/private/laravel' . '/../../web';` but, clearly, that assumption was invalid and unfounded.

In the interest of not besmirching anyone's good name, perhaps there's no reason to mention the other library. This was a classic case of "operator error" it seems.

**Thank you** for taking the time to read through the details, understand the issue, and point me straight towards a proper solution. Your assistance is deeply appreciated.

:thumbsup:

**Comment by stof** [comment]
@mmic-bjohnson well, a library having an impact on the composer autoloading generation when changing its config is worth knowing, to be able to help future people having a similar issue.
The goal is not to blame the library, but to be able to help more efficiently people facing the same operator error than you in the future.

**Comment by mmic-bjohnson** [comment]
@stof Sure, that makes sense. I've already shouldered the blame, so, what's the harm? :)

While I'm not positive *which* Cartalyst library is responsible for constructing the PSR-4 autoloader paths, it's possibly `cartalyst/platform-installer` (it's subscription-based, and thus Private).

As a bit of background, my environment forces me to have two different paths for "private" resources and "public" resources (those that should be accessible via HTTP[S]), and the public path is not a child of the private path. This wreaks absolute havoc in Laravel, which, apparently, is not designed allow for this type of configuration -- at least not without pulling a few teeth (see: https://laracasts.com/discuss/channels/general-discussion/where-do-you-set-public-directory-laravel-5 ).

Hope this helps!
